scala/scala-impl/src/org/jetbrains/plugins/scala/lang/psi/implicits/ImplicitProcessor.scala

```diff
@@ -206,24 +206,55 @@ object ImplicitProcessor {
       }
     }
 
+    // Java Raw types are converted to F[ScExistentialArgument.Deferred("A", .....), ...]
+    // In combination with F-Bounds this can lead to different instantiations that are not ==,
+    // but would not reveal further parts of the type.
+    //
+    // Here, we convert such existential arguments to stand-in types that have a useful
+    // equals/hashCode implementation, and use this as the marker in the `visitedType` set.
+    def convertRawArgs(tp: ScType): ScType = {
+      def rawArgToDummy(tp: ScType) = tp match {
+        case existentialArgument: ScExistentialArgument =>
+          existentialArgument.typeParamOfRawArg match {
+            case Some(typeParam) =>
+              ScAbstractType(typeParam, existentialArgument.lower, existentialArgument.upper)
+            case None =>
+              tp
+          }
+        case tp => tp
+      }
+      def isRawArg(tp: ScType) = tp match {
+        case existentialArgument: ScExistentialArgument =>
+          existentialArgument.typeParamOfRawArg.isDefined
+        case _ => false
+      }
+      tp match {
+        case ParameterizedType(des, targs) =>
+          if (targs.exists(isRawArg)) {
+            val targs1 = targs.map(rawArgToDummy)
+            ScParameterizedType(des, targs1)
+          } else tp
+        case _ => tp
+      }
+    }
+
     def collectParts(tp: ScType): Unit = {
       ProgressManager.checkCanceled()
-      if (visited.contains(tp)) return
```

Comment:
Not a functional change, this is a micro-optimization to avoid double calculation of `tp.hashCode` and the hash bucket search.